tests/AutoFixture.TUnit.Tests/Internal/DataSourceTests.cs

```diff
@@ -0,0 +1,111 @@
+﻿using AutoFixture.TUnit.Internal;
+using AutoFixture.TUnit.Tests.TestTypes;
+using TUnit.Assertions.AssertConditions.Throws;
+
+namespace AutoFixture.TUnit.Tests.Internal;
+
+public class DataSourceTests
+{
+    [Test]
+    public async Task SutIsTestDataSource()
+    {
+        // Arrange
+        var sut = new DelegatingDataSource();
+
+        // Assert
+        await Assert.That(sut).IsAssignableTo<IDataSource>();
+    }
+
+    [Test]
+    public async Task ReturnSingleEmptyArrayWhenMethodHasNoParameters()
+    {
+        // Arrange
+        var sut = new DelegatingDataSource();
+        var testMethod = typeof(SampleTestType)
+            .GetMethod(nameof(SampleTestType.TestMethodWithoutParameters));
+
+        // Act
+        var result = sut.GenerateDataSources(DataGeneratorMetadataHelper.CreateDataGeneratorMetadata(testMethod)).ToArray();
+
+        // Assert
+        var item = await Assert.That(result).HasSingleItem();
+
+        await Assert.That(item).IsEmpty();
+    }
+
+    [Test]
+    public async Task ThrowsWhenNoDataFoundForMethod()
+    {
+        // Arrange
+        var sut = new DelegatingDataSource { TestData = null! };
+        var testMethod = typeof(SampleTestType)
+            .GetMethod(nameof(SampleTestType.TestMethodWithSingleParameter));
+
+        // Act & Assert
+        await Assert.That(() => sut.GenerateDataSources(DataGeneratorMetadataHelper.CreateDataGeneratorMetadata(testMethod)).ToArray()).ThrowsExactly<InvalidOperationException>();
+    }
+
+    [Test]
+    public async Task ReturnSingleArrayWithSingleItemWhenMethodHasSingleParameter()
+    {
+        // Arrange
+        var sut = new DelegatingDataSource
+        {
+            TestData = [["hello"]]
+        };
+        var testMethod = typeof(SampleTestType)
+            .GetMethod(nameof(SampleTestType.TestMethodWithSingleParameter));
+
+        // Act
+        var result = sut.GenerateDataSources(DataGeneratorMetadataHelper.CreateDataGeneratorMetadata(testMethod)).ToArray();
+
+        // Assert
+        var testData = await Assert.That(result).HasSingleItem();
```

Comment:
With nullability enabled this section does not build.
In this particular case `testData` is a nullable `Func<>` and the next assert uses a `Then` overload that accepts only non-nullable functions.

What is the recommended way to write asserts in this case? Do I implement the function explicitly to check for null and then pass a non-nullable function result to the next assert, or should I ignore it and let `ValueDelegateAssertionBuilder` throw a null reference exception?

Reply:
If you do `Assert.That(...).IsNotNull().And.OtherAssertions...` then the nullability issues should go away